sql/tests: TestMonotonicInserts/distsql=on failed

sql/tests.TestMonotonicInserts/distsql=on [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3200931&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3200931&tab=artifacts#/) on master @ [9f39af35fd81a7754e6123862b63247e3277b458](https://github.com/cockroachdb/cockroach/commits/9f39af35fd81a7754e6123862b63247e3277b458):


```
Slow failing tests:
TestMonotonicInserts/distsql=on - 2317.81s

Slow passing tests:
Example_demo - 210.29s
TestMergeQueueSeesNonVoters - 195.92s
TestExternalSortRandomized - 158.21s
TestAggregatorRandom - 143.75s
TestSortRandomized - 112.27s
TestIndexCleanupAfterAlterFromRegionalByRow - 110.84s
TestExternalSortMemoryAccounting - 104.40s
TestUnorderedDistinctRandom - 99.64s
TestLearnerSnapshotFailsRollback - 97.97s
TestBackupRestoreTenant - 94.55s
TestImportAvro - 83.87s
TestRestoreOldVersions - 77.22s
TestInvertedIndex - 70.65s
TestExternalDistinctSpilling - 66.81s
TestStoreRangeMergeTimestampCache - 66.38s
TestChangefeedSchemaChangeAllowBackfill - 58.81s
TestDefaultColumns - 56.85s
TestReadWriteFile - 56.74s
TestScatterRandomizeLeases - 56.47s
TestMultiNodeExportStmt - 56.22s
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestMonotonicInserts/distsql=on PKG=./pkg/sql/tests TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
/cc @cockroachdb/sql-queries petermattis
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMonotonicInserts/distsql=on.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

I've stressed it with `make stress PKG=./pkg/sql/tests TESTS=TestMonotonicInserts TESTFLAGS='-test.v -show-logs' TESTTIMEOUT=30s STRESSFLAGS='-p 1'`, and it timed out pretty quickly (on 12th run, at 3 minute mark). And to me it looks like a deadlock, block like this one is repeated a couple of thousand times:
```
I210720 16:51:15.662527 20124 sql/tests/monotonic_insert_test.go:134  [-] 11005  363.002: attempt 11
I210720 16:51:15.662752 20124 sql/tests/monotonic_insert_test.go:134  [-] 11006  363.002: read max val
I210720 16:51:15.664405 20124 sql/tests/monotonic_insert_test.go:134  [-] 11007  363.002: read max row for val=368
I210720 16:51:15.666129 20124 sql/tests/monotonic_insert_test.go:134  [-] 11008  363.002: insert
I210720 16:51:15.666918 20124 sql/tests/monotonic_insert_test.go:134  [-] 11009  363.002: commit
```
...
```
I210720 16:51:35.254051 20124 sql/tests/monotonic_insert_test.go:134  [-] 26750  363.002: attempt 3158
I210720 16:51:35.254224 20124 sql/tests/monotonic_insert_test.go:134  [-] 26751  363.002: read max val
I210720 16:51:35.257515 20124 sql/tests/monotonic_insert_test.go:134  [-] 26752  363.002: read max row for val=368
I210720 16:51:35.260935 20124 sql/tests/monotonic_insert_test.go:134  [-] 26753  363.002: insert
I210720 16:51:35.261624 20124 sql/tests/monotonic_insert_test.go:134  [-] 26754  363.002: commit
I210720 16:51:35.263058 20124 sql/tests/monotonic_insert_test.go:134  [-] 26755  363.002: attempt 3159
I210720 16:51:35.263238 20124 sql/tests/monotonic_insert_test.go:134  [-] 26756  363.002: read max val
I210720 16:51:35.266559 20124 sql/tests/monotonic_insert_test.go:134  [-] 26757  363.002: read max row for val=368
I210720 16:51:35.270079 20124 sql/tests/monotonic_insert_test.go:134  [-] 26758  363.002: insert
panic: test timed out after 30s
```
Seems like a KV issue.

--- a/pkg/kv/kvclient/kvcoord/BUILD.bazel
+++ b/pkg/kv/kvclient/kvcoord/BUILD.bazel
@@ -148,6 +148,7 @@ go_test(
         "//pkg/kv/kvbase",
         "//pkg/kv/kvclient/rangecache:with-mocks",
         "//pkg/kv/kvserver",
+        "//pkg/kv/kvserver/closedts",
         "//pkg/kv/kvserver/concurrency/lock",
         "//pkg/kv/kvserver/kvserverbase",
         "//pkg/kv/kvserver/tscache",

--- a/pkg/kv/kvclient/kvcoord/txn_test.go
+++ b/pkg/kv/kvclient/kvcoord/txn_test.go
@@ -20,6 +20,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/keys"
 	"github.com/cockroachdb/cockroach/pkg/kv"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/closedts"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/concurrency/lock"
 	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/tscache"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
@@ -895,4 +896,54 @@ func TestTxnReturnsWriteTooOldErrorOnConflictingDeleteRange(t *testing.T) {
 	require.NotNil(t, err)
 	require.Regexp(t, "TransactionRetryWithProtoRefreshError: WriteTooOldError", err)
 	require.Len(t, b.Results[0].Keys, 0)
+}
+
+// TestRetrySerializableBumpsToNow verifies that transaction read time is forwarded to the
+// current HLC time rather than closed timestamp to give it enough time to retry.
+// To achieve that, test fixes transaction time to prevent refresh from succeeding first
+// then waits for closed timestamp to advance sufficiently and commits. Since write
+// can't succeed below closed timestamp and commit timestamp has leaked and can't be bumped
+// txn is restarted. Test then verifies that it can proceed even if closed timestamp
+// is again updated.
+func TestRetrySerializableBumpsToNow(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	s := createTestDB(t)
+	defer s.Stop()
+	ctx := context.Background()
+
+	bumpClosedTimestamp := func(delay time.Duration) {
+		s.Manual.Increment(delay.Nanoseconds())
+		// We need to bump closed timestamp for clock increment to have effect
+		// on further kv writes. Putting anything into proposal buffer will
+		// trigger achieve this.
+		// We write a value to a non-overlapping key to ensure we are not
+		// bumping test transaction. We can't use reads because they could
+		// only bump tscache directly which we try not to do in this test case.
+		require.NoError(t, s.DB.Put(ctx, roachpb.Key("z"), []byte{0}))
+	}
+
+	attempt := 0
+	require.NoError(t, s.DB.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+		require.Less(t, attempt, 2, "Transaction can't recover after initial retry error, too many retries")
+		closedTsTargetDuration := closedts.TargetDuration.Get(&s.Cfg.Settings.SV)
+		delay := closedTsTargetDuration / 2
+		if attempt == 0 {
+			delay = closedTsTargetDuration * 2
+		}
+		bumpClosedTimestamp(delay)
+		attempt++
+		// Fixing transaction commit timestamp to disallow read refresh.
+		_ = txn.CommitTimestamp()
+		// Perform a scan to populate the transaction's read spans and mandate a refresh
+		// if the transaction's write timestamp is ever bumped. Because we fixed the
+		// transaction's commit timestamp, it will be forced to retry.
+		_, err := txn.Scan(ctx, roachpb.Key("a"), roachpb.Key("p"), 1000)
+		require.NoError(t, err, "Failed Scan request")
+		// Perform a write, which will run into the closed timestamp and get pushed.
+		require.NoError(t, txn.Put(ctx, roachpb.Key("b"), []byte(fmt.Sprintf("value-%d", attempt))))
+		return nil
+	}))
+	require.Greater(t, attempt, 1, "Transaction is expected to retry once")
 }

--- a/pkg/roachpb/data.go
+++ b/pkg/roachpb/data.go
@@ -1457,9 +1457,28 @@ func PrepareTransactionForRetry(
 		txn.WriteTimestamp.Forward(tErr.PusheeTxn.WriteTimestamp)
 		txn.UpgradePriority(tErr.PusheeTxn.Priority - 1)
 	case *TransactionRetryError:
-		// Nothing to do. Transaction.Timestamp has already been forwarded to be
-		// ahead of any timestamp cache entries or newer versions which caused
-		// the restart.
+		// Transaction.Timestamp has already been forwarded to be ahead of any
+		// timestamp cache entries or newer versions which caused the restart.
+		if tErr.Reason == RETRY_SERIALIZABLE {
+			// For RETRY_SERIALIZABLE case, we want to bump timestamp further than
+			// timestamp cache.
+			// This helps transactions that had their commit timestamp fixed (See
+			// roachpb.Transaction.CommitTimestampFixed for details on when it happens)
+			// or transactions that hit read-write contention and can't bump
+			// read timestamp because of later writes.
+			// Upon retry, we want those transactions to restart on now() instead of
+			// closed ts to give them some time to complete without a need to refresh
+			// read spans yet again and possibly fail.
+			// The tradeoff here is that transactions that failed because they were
+			// waiting on locks or were slowed down in their first epoch for any other
+			// reason (e.g. lease transfers, network congestion, node failure, etc.)
+			// would have a chance to retry and succeed, but transactions that are
+			// just slow would still retry indefinitely and delay transactions that
+			// try to write to the keys this transaction reads because reads are not
+			// in the past anymore.
+			now := clock.Now()
+			txn.WriteTimestamp.Forward(now)
+		}
 	case *WriteTooOldError:
 		// Increase the timestamp to the ts at which we've actually written.
 		txn.WriteTimestamp.Forward(writeTooOldRetryTimestamp(tErr))

--- a/pkg/sql/tests/monotonic_insert_test.go
+++ b/pkg/sql/tests/monotonic_insert_test.go
@@ -77,7 +77,6 @@ type mtClient struct {
 //   https://github.com/jepsen-io/jepsen/blob/master/cockroachdb/src/jepsen/cockroach/monotonic.clj
 func TestMonotonicInserts(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.WithIssue(t, 67802, "flaky test")
 
 	for _, distSQLMode := range []sessiondatapb.DistSQLExecMode{
 		sessiondatapb.DistSQLOff, sessiondatapb.DistSQLOn,

